File: app/src/main/java/com/apisod/inventoryar/application/GlobalApplication.java
//package com.apisod.inventoryar.application;
//
//import android.app.Application;
//import android.content.Context;
//import android.content.SharedPreferences;
//
//import androidx.multidex.MultiDex;
//
//import com.apisod.inventoryar.constants.Constants;
//import com.google.firebase.FirebaseApp;
//import com.google.firebase.firestore.FirebaseFirestore;
//import com.google.firebase.firestore.FirebaseFirestoreSettings;
//import com.google.firebase.storage.FirebaseStorage;
//
//import dagger.hilt.android.HiltAndroidApp;
//
//@HiltAndroidApp
//public class GlobalApplication extends Application {
//
//    private static GlobalApplication globalApplication;
//    private static FirebaseFirestore firebaseFirestore;
//    private static FirebaseStorage storage;
//    private static SharedPreferences mPrefs;
//
//    public static GlobalApplication getInstance() {
////        if (globalApplication == null) {
////            System.out.println("GlobalApplication new getInstance");
////            globalApplication = new GlobalApplication();
////        }
//        return globalApplication;
//    }
//
//    @Override
//    protected void attachBaseContext(Context base) {
//        super.attachBaseContext(base);
//        MultiDex.install(this);
//    }
//
//    @Override
//    public void onCreate() {
//        super.onCreate();
//        globalApplication = this;
//
//        try {
//            FirebaseApp.initializeApp(this);
//        } catch (Exception e) {
//            e.printStackTrace();
//        }
//
//        initFirebaseFirestore();
//    }
//
//    public static FirebaseFirestore getFirebaseFirestore() {
//        return firebaseFirestore;
//    }
//
//    private void initFirebaseFirestore() {
//        try {
//            if (firebaseFirestore == null) {
//                firebaseFirestore = FirebaseFirestore.getInstance();
//                FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
//                        .setPersistenceEnabled(true)
//                        .build();
//                firebaseFirestore.setFirestoreSettings(settings);
//            }
//        } catch (Exception e) {
//            e.printStackTrace();
//        }
//    }
//
//    public static FirebaseStorage getFirebaseStorage() {
//        if (storage == null)
//            storage = FirebaseStorage.getInstance();
//        return storage;
//    }
//
//    public static SharedPreferences getSharedPrefs(Context context) {
//        if (mPrefs == null)
//            mPrefs = context.getSharedPreferences(Constants.MYPREF, Context.MODE_PRIVATE);
//        return mPrefs;
//    }
//}

File: app/src/main/java/com/apisod/inventoryar/application/GlobalApplication.kt
package com.apisod.inventoryar.application

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import androidx.multidex.MultiDex
import com.apisod.inventoryar.constants.Constants
import com.google.firebase.FirebaseApp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.google.firebase.storage.FirebaseStorage
import dagger.hilt.android.HiltAndroidApp

@HiltAndroidApp
class GlobalApplication :
    Application() {

    companion object {

        private lateinit var instance:
                GlobalApplication

        private lateinit var firebaseFirestore:
                FirebaseFirestore

        private lateinit var firebaseStorage:
                FirebaseStorage

        private var sharedPreferences:
                SharedPreferences? = null

        fun getInstance():
                GlobalApplication {

            return instance
        }

        fun provideFirebaseFirestore():
                FirebaseFirestore {

            return firebaseFirestore
        }

        fun provideFirebaseStorage():
                FirebaseStorage {

            return firebaseStorage
        }

        fun getSharedPrefs(
            context: Context
        ): SharedPreferences {

            if (sharedPreferences == null) {

                sharedPreferences =
                    context.getSharedPreferences(
                        Constants.MYPREF,
                        Context.MODE_PRIVATE
                    )
            }

            return sharedPreferences!!
        }
    }

    override fun attachBaseContext(
        base: Context
    ) {

        super.attachBaseContext(base)

        MultiDex.install(this)
    }

    override fun onCreate() {

        super.onCreate()

        instance = this

        initializeFirebase()
    }

    private fun initializeFirebase() {

        try {

            FirebaseApp.initializeApp(
                this
            )

            firebaseFirestore =
                FirebaseFirestore
                    .getInstance()

            val settings =
                FirebaseFirestoreSettings
                    .Builder()
                    .setPersistenceEnabled(true)
                    .build()

            firebaseFirestore
                .firestoreSettings =
                settings

            firebaseStorage =
                FirebaseStorage
                    .getInstance()

        } catch (e: Exception) {

            e.printStackTrace()
        }
    }
}